Can not assign dict whos key is not string as module attribute

Hi,
It seems current flax.linen not allow assign a dict with non-string keys to the module attribute.
See below simple example:
it will trigger error:
`AssertionError: A state dict must only have string keys.`

Questions:
1. Is it an intended behavior? Why?
2. If it is intended, is there any work around? As it is quite possible we need assign the information contained by a dict to the module, and the key of the dict may not string.

```python

import flax.linen as nn
import jax
import jax.numpy as jnp


class Foo(nn.Module):
    def setup(self):
        self.a = {(1, 2): 3} # here the dict using tuple as key

    @nn.compact
    def __call__(self, x):
        return x


foo = Foo()
rng = jax.random.PRNGKey(0)
x = jnp.ones(shape=(3, 3))
vars = foo.init({"params": rng}, x)
out = foo.apply(vars, x)
print(out)
```

--- a/flax/serialization.py
+++ b/flax/serialization.py
@@ -117,11 +117,14 @@ def _restore_list(xs, state_dict: Dict[str, Any]) -> List[Any]:
 
 
 def _dict_state_dict(xs: Dict[str, Any]) -> Dict[str, Any]:
-  return {key: to_state_dict(value) for key, value in xs.items()}
+  str_keys = set(str(k) for k in xs.keys())
+  if len(str_keys) != len(xs):
+    raise ValueError(f'Dict keys do not have a unique string representation: {str_keys}')
+  return {str(key): to_state_dict(value) for key, value in xs.items()}
 
 
 def _restore_dict(xs, states: Dict[str, Any]) -> Dict[str, Any]:
-  return {key: from_state_dict(value, states[key])
+  return {key: from_state_dict(value, states[str(key)])
           for key, value in xs.items()}
 
 

--- a/tests/linen/linen_module_test.py
+++ b/tests/linen/linen_module_test.py
@@ -159,6 +159,22 @@ class ModuleTest(absltest.TestCase):
       {'lyrs1_a': {'kernel': (10, 3)},
       'lyrs1_b': {'kernel': (3, 3)}})
 
+  def test_setup_dict_nonstring_keys(self):
+    class Foo(nn.Module):
+      def setup(self):
+          self.a = {(1, 2): nn.Dense(2)} # here the dict using tuple as key
+
+      @nn.compact
+      def __call__(self, x):
+          return self.a[(1, 2)](x)
+
+    foo = Foo()
+    x = jnp.ones(shape=(1, 3))
+    params = foo.init(random.PRNGKey(0), x)['params']
+    param_shape = jax.tree_map(jnp.shape, params)
+    self.assertEqual(param_shape,
+      {'a_(1, 2)': {'kernel': (3, 2), 'bias': (2,)}})
+
   def test_setup_cloning(self):
     class MLP(nn.Module):
       def setup(self):

